# CassetteCrafter
CassetteCrafter automates the creation of DNA sequence libraries containing all combinations of desired mutations, formatted for Golden Gate Assembly workflows.

## Installation and Setup

### Requirements

- Python 3.x
- Flask 
- BioPython
- Pandas
- Optional: Pytest
  
### Setup

1. Clone the repository:  
   `git clone <https://github.com/yourusername/cassettecrafter.git>`

2. Create the conda environment containing packages needed to run the software:   
 `conda env create -f environment.yml`

3. Activate the conda environment:  
   `conda activate cassettecrafter`

## Usage
From the src directory, sequences can be generated by running main.py

Usage:  `python3 main.py -u [allowed mutations file] -e [restriction enzyme name] -m [minimum oligo size] -M [maximum oligo size] -f [file with starting gene sequence]`  
or  
`python3 main.py --mutations [allowed mutations file] --enzyme_name [restriction enzyme name] --min_oligo_size [minimum oligo size] --max_oligo_size [maximum oligo size] --gene_file [file with starting gene sequence]`

To run an example from the `cassettecrafter` directory (parent directory of `src`):`python3 main.py -m ./test_data/demo_mutation_list.csv -f ./test_data/LY011_test_seq_single.csv`

### Proper formatting of input:
* gene_file : Gene sequence (currently a codon sequence) can be provided as a fasta file in the traditional format or a csv without a header, with each row formatted as [gene_name],[DNA_sequence]
    * cassettecrafter/test_data/LY011_test_seq_single.fa
    * cassettecrafter/test_data/LY011_test_seq_single.csv
* mutations : Allowed mutations can currently only be provided as a csv, with a header row. The first column should correspond to the position in the amino acid sequence (1-indexed), and the second column should contain a string of allowed amino acids mutations at that positions, separated by commas.  
  
  **Example Formatting:**
  
  | Position | Amino Acids |
  |----------|-------------|
  | 4        | E, H, Y, D  |
  | 16       | A, G, P     |
  | 27       | E, R, H     |

**File Location:**  
- `cassettecrafter/test_data/demo_mutation_list.csv`

### Enzyme Data
* Information about restriction enzymes used for Golden Gate Assembly is contained in a csv file. Can contain data about multiple enzymes, but only one will be used for inserting recognition and cut sites into the provided gene, based on the user-provided enzyme name. The file available from this repository has information for a number of common enzymes, but users can specify additional enzyme information if necessary.
* Information should be provided as a csv with a header row, and the following columns:  
  Enzyme, Fwd_recognition_site, Rev_recognition_site, spacer_length, OH_length  


    **Example Formatting:**
  
|  Enzyme | Fwd_recognition_site | Rev_recognition_site | Spacer Length | OH Length |
|---------|----------------------|----------------------|---------------|-----------|
| BbsI    | GAAGAC               | GTCTTC               | 2             | 4         |
| PaqCI   | CACCTGC              | GTGGACG              | 4             | 4         |

- spacer_length: The number of bases between the recognition sites and the start of the cut site.
-  OH_length: The length of the overhang generated by the enzyme cut.
  
**File Location:**  
- `cassettecrafter/test_data/enzyme_sites.csv`
     


